core/src/test/java/io/parsingdata/metal/data/selection/ByTokenTest.java

```diff
@@ -229,4 +250,35 @@ public void getSubRef() throws IOException {
         assertTrue(refs.tail.head.isGraph());
     }
 
+    private final Token subSeq = seq(any("b"), any("c"));
+
+    @Test
+    public void getAllRootsSingle() throws IOException {
+        final Token topSeq = seq(any("a"), subSeq);
+        final ParseResult result = topSeq.parse(stream(1, 2, 3), enc());
+        assertTrue(result.succeeded);
+        final ParseItemList seqs = ByToken.getAllRoots(result.environment.order, subSeq);
+        assertEquals(1, seqs.size);
+        assertEquals(subSeq, seqs.head.getDefinition());
+        final ParseValue c = seqs.head.asGraph().head.asValue();
+        assertEquals(3, c.asNumeric().intValue());
+        assertEquals(2, c.getOffset());
+    }
+
+    @Test
+    public void getAllRootsMulti() throws IOException {
+        final Token topSeq = seq(any("a"), subSeq, subSeq);
+        final ParseResult result = topSeq.parse(stream(1, 2, 3, 2, 3), enc());
+        assertTrue(result.succeeded);
+        final ParseItemList seqs = ByToken.getAllRoots(result.environment.order, subSeq);
+        assertEquals(2, seqs.size);
+        assertEquals(subSeq, seqs.head.getDefinition());
+        assertEquals(subSeq, seqs.tail.head.getDefinition());
+        final ParseValue c1 = seqs.head.asGraph().head.asValue();
+        assertEquals(3, c1.asNumeric().intValue());
+        final ParseValue c2 = seqs.tail.head.asGraph().head.asValue();
+        assertEquals(3, c2.asNumeric().intValue());
+        assertNotEquals(seqs.head.asGraph().head, seqs.tail.head.asGraph().head);
+    }
+
```

Comment:
I've added a `Sub`-test but not one that uses self-referencing, mainly because that causes a problem: the `definition` of a `Token` contained in a custom `Token` (anonymous or not) actually doesn't point to the outside class, but the embedded `Token`, which is difficult to access and reference.